vowpalwabbit/cb_explore_adf_rnd.cc

```diff
@@ -0,0 +1,307 @@
+#include "reductions.h"
+#include "cb_adf.h"
+#include "rand48.h"
+#include "bs.h"
+#include "gd_predict.h"
+#include "gen_cs_example.h"
+#include "cb_explore.h"
+#include "explore.h"
+#include <vector>
+#include <algorithm>
+#include <cmath>
+
+namespace VW
+{
+namespace cb_explore_adf
+{
+namespace rnd
+{
+struct cb_explore_adf_rnd
+{
+ private:
+  float epsilon;
+  float alpha;
+  float sqrtinvlambda;
+  uint32_t numrnd;
+
+  size_t increment;
+  vw* all;
+
+ public:
+  cb_explore_adf_rnd(float _epsilon, float _alpha, float _invlambda, size_t _numrnd, size_t _increment, vw* _all) : epsilon(_epsilon), alpha(_alpha), sqrtinvlambda(std::sqrt(_invlambda)), numrnd(_numrnd), increment(_increment), all(_all) { }
+  ~cb_explore_adf_rnd() = default;
+
+  // Should be called through cb_explore_adf_base for pre/post-processing
+  void predict(VW::LEARNER::multi_learner& base, multi_ex& examples) { predict_or_learn_impl<false>(base, examples); }
+  void learn(VW::LEARNER::multi_learner& base, multi_ex& examples) { predict_or_learn_impl<true>(base, examples); }
+
+ private:
+  template <bool is_learn>
+  void predict_or_learn_impl(VW::LEARNER::multi_learner& base, multi_ex& examples);
+
+  std::vector<float> bonuses;
+  std::vector<float> initials;
+  float get_initial_prediction(example*);
+  void get_initial_predictions(multi_ex&, uint32_t);
+  void zero_bonuses(multi_ex&);
+  void accumulate_bonuses(multi_ex&);
+  void finish_bonuses();
+  void compute_ci(v_array<ACTION_SCORE::action_score>&, float);
+
+  CB::cb_class save_class;
+  template<bool> void save_labels(multi_ex&);
+  template<bool> void make_fake_rnd_labels(multi_ex&);
+  template<bool> void restore_labels(multi_ex&);
+  template<bool> void base_learn_or_predict(VW::LEARNER::multi_learner&, multi_ex&, uint32_t);
+};
+
+void cb_explore_adf_rnd::zero_bonuses(multi_ex& examples)
+{
+  bonuses.clear();
+  bonuses.reserve(examples.size());
+  for (auto* ec: examples) { (void) ec; bonuses.push_back(0.f); }
+}
+
+void cb_explore_adf_rnd::accumulate_bonuses(multi_ex& examples)
+{
+  v_array<ACTION_SCORE::action_score>& preds = examples[0]->pred.a_s;
+  for (const auto& p : preds)
+    {
+      float score = p.score - initials[p.action];
+      bonuses[p.action] += score * score;
+    }
+}
+
+void cb_explore_adf_rnd::finish_bonuses()
+{
+  for (auto& b: bonuses) { b = sqrt(b / numrnd); }
+}
+
+void cb_explore_adf_rnd::compute_ci(v_array<ACTION_SCORE::action_score>& preds,
+                                    float maxbonus)
+{
+  const float eulergamma = 0.57721566490153286;
+  for (auto& p: preds) { p.score -= eulergamma * (bonuses[p.action] - maxbonus); }
+}
+
+namespace
+{
+  bool is_the_labeled_example(const example* ec)
+    {
+      return (   ec->l.cb.costs.size() == 1
+              && ec->l.cb.costs[0].cost != FLT_MAX
+              && ec->l.cb.costs[0].probability > 0);
+    }
+}
+
+template <bool is_learn>
+void cb_explore_adf_rnd::save_labels(multi_ex& examples)
+{
+  if (is_learn)
```

Comment:
Given that several of these functions do nothing in the is_learn = false case does it make sense to uplift this check before calling. It's a little misleading to have the function be a noop in my mind

Reply:
This way made the logic from 223-234 more readable.  Other ways involved sprinkling conditionals and/or duplicating code blocks.

Reply:
Yeah I see what you mean. I thought about that when I saw it too. Makes sense.